<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <!-- topic parameters -->
  <arg name="estimated_pose_topic"      default="/mavros/vision_pose/pose" />
  <arg name="uwb_range_topic"           default="/uwb_driver_node/uwb" />
  <arg name="start_service"             default="/uwb_init/start" />
  <arg name="reset_service"             default="/uwb_init/reset" />
  <arg name="initialization_service"    default="/uwb_init/initialize" />
  <arg name="refine_service"            default="/uwb_init/refine" />
  <arg name="logger_level"              default="info" />

  <!-- calibration -->
  <arg name="calib_p_UinI"              default="[0.03, -0.04, -0.03]" />

  <!-- node parameters -->
  <arg name="method"                    default="double" />
  <arg name="bias_type"                 default="constant" />
  
  <!-- ls solver parameters -->
  <arg name="position_std"              default="0.03" />
  <arg name="uwb_range_std"             default="0.1" />

  <!-- nls solver parameters -->
  <arg name="step_norm_stop_condition"            	default="0.00001" />
  <arg name="residual_mse_stop_condition"         	default="0.01" />
  <arg name="max_iterations"            		default="10000" />
  <arg name="step_vector"   default="[0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1.0, 2.0, 5.0, 10.0, 20.0, 50.0, 100.0]" />

  <node name="uwb_init" pkg="uwb_init_ros" type="uwb_init" args="" output="screen" respawn="false">

    <!-- topics remaps -->
    <param name="estimated_pose_topic"      type="str"  value="$(arg estimated_pose_topic)" />
    <param name="uwb_range_topic"           type="str"  value="$(arg uwb_range_topic)" />
    <param name="start_service"             type="str"  value="$(arg start_service)" />
    <param name="reset_service"             type="str"  value="$(arg reset_service)" />
    <param name="initialization_service"    type="str"  value="$(arg initialization_service)" />
    <param name="refine_service"            type="str"  value="$(arg refine_service)" />
    <param name="logger_level"              type="str"  value="$(arg logger_level)" />

    <!-- calibration -->
    <rosparam param="p_ItoU"                subst_value="True">$(arg calib_p_UinI)</rosparam>

    <!-- node parameters -->
    <param name="method"                    type="str"  value="$(arg method)" />
    <param name="bias_type"                 type="str"  value="$(arg bias_type)" />

    <!-- ls solver parameters -->
    <param name="position_std"              type="double"  value="$(arg position_std)" />
    <param name="uwb_range_std"             type="double"  value="$(arg uwb_range_std)" />

    <!-- nls solver parameters -->
    <param name="step_norm_stop_condition"          type="double"  value="$(arg step_norm_stop_condition)" />
    <param name="residual_mse_stop_condition"       type="double"  value="$(arg residual_mse_stop_condition)" />
    <param name="max_iterations"                    type="int"  value="$(arg max_iterations)" />
    <rosparam param="step_vec"                      subst_value="True">$(arg step_vector)</rosparam>

  </node>

</launch>
